consensus/ibft/signer/ecdsa.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/0xPolygon/polygon-edge/consensus/ibft/validators"
+	"github.com/0xPolygon/polygon-edge/crypto"
+	"github.com/0xPolygon/polygon-edge/secrets"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/umbracle/fastrlp"
+)
+
+type ECDSAKeyManager struct {
+	key     *ecdsa.PrivateKey
+	address types.Address
+}
+
+type SerializedSeal [][]byte
+
+func NewECDSAKeyManager(manager secrets.SecretsManager) (KeyManager, error) {
+	key, err := obtainOrCreateECDSAKey(manager)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewECDSAKeyManagerFromKey(key), nil
+}
+
+func NewECDSAKeyManagerFromKey(key *ecdsa.PrivateKey) KeyManager {
+	return &ECDSAKeyManager{
+		key:     key,
+		address: crypto.PubKeyToAddress(&key.PublicKey),
+	}
+}
+
+func NewMockECDSAKeyManager(addr types.Address) KeyManager {
+	return &ECDSAKeyManager{
+		address: addr,
+	}
+}
+
+func (s *ECDSAKeyManager) Address() types.Address {
+	return s.address
+}
+
+func (s *ECDSAKeyManager) NewEmptyIstanbulExtra() *IstanbulExtra {
+	return &IstanbulExtra{
+		Validators:          &validators.ECDSAValidatorSet{},
+		CommittedSeal:       &SerializedSeal{},
+		ParentCommittedSeal: &SerializedSeal{},
+	}
+}
+
+func (s *ECDSAKeyManager) NewEmptyCommittedSeal() Sealer {
+	return &SerializedSeal{}
+}
+
+func (s *ECDSAKeyManager) SignSeal(data []byte) ([]byte, error) {
+	return crypto.Sign(s.key, data)
+}
+
+func (s *ECDSAKeyManager) SignCommittedSeal(data []byte) ([]byte, error) {
+	return crypto.Sign(s.key, data)
+}
+
+func (s *ECDSAKeyManager) Ecrecover(sig, digest []byte) (types.Address, error) {
+	return ecrecoverImpl(sig, digest)
+}
+
+func (s *ECDSAKeyManager) GenerateCommittedSeals(
+	sealMap map[types.Address][]byte,
+	extra *IstanbulExtra,
+) (Sealer, error) {
+	seals := [][]byte{}
+
+	for _, seal := range sealMap {
+		if len(seal) != IstanbulExtraSeal {
+			return nil, ErrInvalidCommittedSealLength
+		}
+
+		seals = append(seals, seal)
+	}
+
+	serializedSeal := SerializedSeal(seals)
+
+	return &serializedSeal, nil
+}
+
+func (s *ECDSAKeyManager) VerifyCommittedSeal(
+	rawCommittedSeal Sealer,
+	digest []byte,
+	rawSet validators.ValidatorSet,
+) (int, error) {
+	committedSeal, ok := rawCommittedSeal.(*SerializedSeal)
+	if !ok {
+		return 0, ErrInvalidCommittedSealType
+	}
+
+	validatorSet, ok := rawSet.(*validators.ECDSAValidatorSet)
+	if !ok {
+		return 0, ErrInvalidValidatorSet
+	}
+
+	return s.verifyCommittedSealsImpl(committedSeal, digest, *validatorSet)
+}
+
+func (s *ECDSAKeyManager) SignIBFTMessage(msg *proto.MessageReq) error {
+	return signMsg(s.key, msg)
+}
+
+func (s *ECDSAKeyManager) ValidateIBFTMessage(msg *proto.MessageReq) error {
+	return ValidateMsg(msg)
+}
+
+func (s *ECDSAKeyManager) verifyCommittedSealsImpl(
+	committedSeal *SerializedSeal,
+	msg []byte,
+	validators validators.ECDSAValidatorSet,
+) (int, error) {
+	numSeals := len(*committedSeal)
+	if numSeals == 0 {
+		return 0, ErrEmptyCommittedSeals
+	}
+
+	visited := map[types.Address]struct{}{}
```

Comment:
What about a map of booleans (set) ?

Reply:
Good idea. Fixed at 52379b35c037308d4ff8e5b77253cacc292214f1